### ISSUE TITLE
[Huawei P20 lite] [AOSP 10 v214] exfat not working.

### ISSUE THREAD

**Issue by ashmonty** [phhusson/treble_experimentations#761]
Micro SD, USB OTG and NFC not working (Using AOSP 10 v206).
Edit: I'm unable to check USB OTG and NFC, but I know that Micro SD still doesn't work (v214).

**Comment by Montelion** [comment]
Micro SD still doesn't work. I can't confirm if USB OTG and NFC work.

**Comment by phhusson** [comment]
Give more details that "doesn't work"

**Comment by Montelion** [comment]
The SD card doesn't show up in any file explorer, and when i boot up the phone, it says "Issue with SanDisk SD card - Tap to fix"
![image](https://user-images.githubusercontent.com/49426949/76767687-49dbfd80-679a-11ea-91d4-5305df463994.png)
If I tap on the notification, the SD formatting screen shows up.
![image](https://user-images.githubusercontent.com/49426949/76767739-595b4680-679a-11ea-8914-70aa81a800ea.png)
If I try to format it it says that the SD card is ready for use, but it still doesn't work. The SD card is not corrupted (I can still use in the custom recovery and on my PC).
Sadly I've lost my USB OTG cable, so I can't check if USB OTG works or not, and I can't check if NFC works or not.

**Comment by phhusson** [comment]
ok, please provide logcat after you insert the sdcard

**Comment by Montelion** [comment]
[logcat_sd.txt](https://github.com/phhusson/treble_experimentations/files/4338937/logcat_sd.txt)

**Comment by phhusson** [comment]
K issue is `03-16 16:31:08.512   399 29828 E vold    : public:179,193 failed to mount /dev/block/vold/public:179,193: Permission denied`
I can't see any denial related to that
I can't understand what kind of issue there could be.
Please give the result of:
`ls -lZ /dev/block/vold/public:179,193` as root

**Comment by Montelion** [comment]
It gives me:
`brw------- 1 root root u:object_r:vold_device:s0 179, 193 2020-03-16 16:34 /dev/block/vold/public:179,193`

**Comment by geppone70** [comment]
I found this:

Disable encryption to use Adoptable Storage (or just because you can)

open /vendor/etc/fstab.hi6250
make a backup just in case
find "/dev/block/bootdevice/by-name/userdata /data f2fs nosuid,nodev,noatime,discard,inlinedata,inlinexatt r wait,check,fileencryption=aes-256-xts:aes-256-cts"
remove ",fileencryption=aes-256-xts:aes-256-cts"
factory reset via stock recovery

Is it related to the issue?

**Comment by Montelion** [comment]
@geppone70 I can't edit the file... I tried mounting /vendor/ read-write on FX, but I still couldn't edit the file. I tried modifying it from TWRP, and I still couldn't do it... Could you please help me?

**Comment by Montelion** [comment]
I also remembered that I was able to remove encryption with Magisk, but the SD still didn't work

**Comment by geppone70** [comment]
i havI still couldn't edit  the file. I'm trying with adb shell but I get always no adb devices in list.

**Comment by geppone70** [comment]
a 32GB sd works, a 64 GB SD don't work, weird

**Comment by phhusson** [comment]
is one formatted as exfat, the other as fat32?

**Comment by geppone70** [comment]
It's fat32

**Comment by phhusson** [comment]
Both of them?

**Comment by geppone70** [comment]
SD 64GB  formatted exFAT don't work
SD 32GB formatted FAT32 works

now I've formatted the SD 64GB FAT32 (using IOMEI partition manager software) and SD 64GB works too

**Comment by phhusson** [comment]
Ok, very good, thanks.
I assume exfat works on stock rom?
Can you give the content of your `/proc/filesystems` ?
(`adb shell cat /proc/filesystems`)

**Comment by Montelion** [comment]
Yes, exfat works on stock rom.
```
nodev   sysfs
nodev   rootfs
nodev   tmpfs
nodev   bdev
nodev   proc
nodev   cpuset
nodev   cgroup
nodev   cgroup2
nodev   devtmpfs
nodev   configfs
nodev   debugfs
nodev   tracefs
nodev   sockfs
nodev   bpf
nodev   pipefs
nodev   ramfs
nodev   devpts
        ext3
        ext2
        ext4
        squashfs
        vfat
        msdos
        exfat
nodev   ecryptfs
nodev   sdcardfs
nodev   cifs
        fuseblk
nodev   fuse
nodev   fusectl
nodev   overlay
        f2fs
nodev   pstore
nodev   selinuxfs
nodev   functionfs
```

**Comment by phhusson** [comment]
`grep exfat /vendor/etc/selinux/*` ?

**Comment by Montelion** [comment]
[grep.txt](https://github.com/phhusson/treble_experimentations/files/4457373/grep.txt)

**Comment by phhusson** [comment]
`echo '(genfscon exfat / (u object_r exfat ((s0) (s0))))' >>
/system/etc/selinux/plat_sepolicy.cil`
(assuming root and rw system)
after that reboot, and tell me if exfat starts working

**Comment by Montelion** [comment]
Nope. I rebooted twice after that, but exfat still isn't working.
The notification says "Unsupported SanDisk SD card", and if i try to format it, it gets stuck at 20%.

**Comment by maxmod** [comment]
Use this formatusb best working https://bit.ly/3oBO5JO

**Comment by montylion** [comment]
> 
> 
> Use this formatusb best working https://bit.ly/3oBO5JO

Thanks for bumping this >1year old issue without adding anything useful lol.
I know how to format an SD card, the issue is that exfat didn't work on the GSI.

While I'm at it I'm gonna close this since I no longer have access to that phone. Feel free to open this back up if the issue is still present on newer versions and/or there have been new developments regarding the issue.

**Comment by chaoscreater** [comment]
I'm having this issue as well.

**Comment by ashmonty** [comment]
@chaoscreater, reopening this! (again, keep in mind I no longer have access to that phone so I won't be able to help)

**Comment by chaoscreater** [comment]
@ashmonty that's all good. I have 2 different phones running on A11 GSI with the exact same issue. I can fix this issue on one of them with a custom kernel that supports exFat, but that kernel is unstable and has some other issues and so ideally looking for a solution on the ROM level.....

I've tried to take the vold file from stock ROM and another A11 custom ROM and replaced it in /system/bin and gave it full permissions. Unfortunately, I couldn't even boot after that point.

There was an old Magisk module that does something with vold and provides exfat support, but that one only works for Oreo unfortunately....

**Comment by soredake** [comment]
Same problem with redmi note 5 pro and latest crdroid 12.1 gsi.

**Comment by soredake** [comment]
> [forum.xda-developers.com/t/request-systemless-filesystem-drivers-for-kernel-mode-and-fuse-support.3807617/#post-79685842](https://forum.xda-developers.com/t/request-systemless-filesystem-drivers-for-kernel-mode-and-fuse-support.3807617/#post-79685842)

This twrp zip is not working for me, and not everyone have twrp so i repacked this twrp zip into magisk module zip, here:
[ntfs-exfat-support.zip](https://github.com/phhusson/treble_experimentations/files/8948418/ntfs-exfat-support.zip)

This module disables selinux in order to mount sdcard as i am not capable (or want to) spend time on learning how to write selinux policies.

**Comment by soredake** [comment]
https://blog.esper.io/android-dessert-bites-27-exfat-on-pixel-532176849/
